File: numerical methods/BisectionMethod.py
#BisectionMethod


#e**x − x**2 + 3x − 2 = 0    ; for 0 ≤ x ≤ 1
#2x cos(2x) − (x + 1)**2 = 0 ; for −3 ≤ x ≤ −2
#x cos x − 2x**2 + 3x − 1 = 0; for 0.2 ≤ x ≤ 0.3

import math as m


l=float(input("Lower Boundary:"))
u=float(input("Upper Boundary:"))


#FUNC. DEFINING

def f(x):
    f = m.e**x -x**2 +3*x - 2                                                 ####### define other f(x)s
    #f=2*x*m.cos(2*x) - (x+1)**2                                              #They'r here just for TestRide
    #f= x * m.cos(x) - 2*x**2 + 3*x -1
    return f


#DHONG

if f(l)*f(u)<0:
    print("There exists at least one root in the closed interval","[",l,",",u,"]")
else:
    print("Ja, Vaag!")


#n(ITTERATION)

d=int(input("result need to be accurate within ? decimal place:"))
t=10**(-d)

n=m.ceil((m.log10(u-l)-m.log10(t))/m.log10(2))
print(" ")
print("We'll need",n,"iterations;")


#MechaParts

for i in range(n):

    p=(u+l)/2

    if f(l)<0:

        if f(p)*f(l)<0:
            u=p
        elif f(p)*f(l)>0:
            l=p

    elif f(l)>0:

        if f(p)*f(l)<0:
            u=p
        elif f(p)*f(l)>0:
            l=p

    elif f(u)<0:

        if f(p)*f(u)<0:
            l=p
        elif f(p)*f(u)>0:
            u=p

    elif f(u)>0:

        if f(p)*f(u)<0:
            l=p
        elif f(p)*f(u)>0:
            u=p




print("And our desired root is:",p)



#/NayeemulHasan.0

File: numerical methods/bisection_method.cpp
#include<bits/stdc++.h>
using namespace std;
double func(double x){
    //replace your desired function here
    return x*x*x+4*x*x-10;
   //return sqrt(x)-cos(x);
}
int main()
{
    double lower,upper,eps,p;

    //Find a interval as small as possible 

    cout<<"Enter the lower and upper bound respectively:"<<endl;
    cin>>lower>>upper;
    cout<<"Enter the precision range:"<<endl;
    cin>>eps;

    if(func(lower)*func(upper)<0){
    while(upper-lower>=eps){
        p=(lower+upper)/2;
        if(func(p)==0)
        break;
        if(func(p)<0)
        lower=p;
        else
        upper=p;
    }
    }
    if(func(lower)*func(upper)>0)
    cout<<"Enter a valid interval using IVP"<<endl;
    
    cout<<setprecision(5)<<"The approximate root is : "<<p<<endl;
   
   
 return 0;
}